### PR TITLE
Make `verify` return a `Result`.

### PR DESCRIPTION
**Pull request by hdevalence**
This also simplifies the verification logic.  Because the verification check
happens in variable time, we don't need to do a constant-time eq check at the
end, so we can drop the `subtle` dependency entirely.

### CODE REVIEW

File: src/ed25519.rs [hunk 1]
```diff
@@ -727,36 +719,27 @@ impl PublicKey {
...
-        ao = self.decompress();
-
-        if ao.is_some() {
-            a = ao.unwrap();
-        } else {
-            return false;
-        }
-        a = -(&a);
+        let A = self.0.decompress()
+            .ok_or_else(|| SignatureError(InternalError::PointDecompressionError))?;
```
Comment (isislovecruft) [inline src/ed25519.rs]: I think I prefer the way I did it with `match` in #31? If that's okay.

File: src/ed25519.rs [hunk 2]
```diff
@@ -694,12 +692,6 @@ impl PublicKey {
         Ok(PublicKey(CompressedEdwardsY(bits)))
     }
 
-    /// Convert this public key to its underlying extended twisted Edwards coordinate.
-    #[inline]
-    fn decompress(&self) -> Option<EdwardsPoint> {
-        self.0.decompress()
-    }
```
Comment (isislovecruft) [inline src/ed25519.rs]: Maybe don't remove this? I'm using it in #31.

File: src/ed25519.rs [hunk 3]
```diff
@@ -727,36 +719,27 @@ impl PublicKey {
...
-        let mut digest: [u8; 64] = [0u8; 64];
-
-        ao = self.decompress();
-
-        if ao.is_some() {
-            a = ao.unwrap();
-        } else {
-            return false;
-        }
-        a = -(&a);
```
Comment (isislovecruft) [inline src/ed25519.rs]: Why did you change to negating `k` instead?

Comment (hdevalence) [reply src/ed25519.rs]: No real reason, I wanted to avoid a point operation, but negating the Edwards point is probably cheaper (since it doesn't require a reduction mod L)

File: src/ed25519.rs (hunk 1 above)
Comment (hdevalence) [reply src/ed25519.rs]: It's up to you, but I think this is a lot cleaner.

File: src/ed25519.rs (hunk 2 above)
Comment (hdevalence) [reply src/ed25519.rs]: I think it's only used there in the same way that it was used here, so my preference would be to remove it there too.

**Comment by hdevalence** [comment]
Changed title to `[WIP]`, I'll update this PR.

What do you think of the error / result changes?

**Comment by isislovecruft** [comment]
The errors and results are good.

File: src/ed25519.rs (hunk 3 above)
Comment (isislovecruft) [reply src/ed25519.rs]: I benched scalar and Edwards point negation and point negation is like twice as fast (they're both in the tens of nanoseconds on my machine so not that it really matters but).

**Comment by hdevalence** [comment]
Rebased onto `develop`.
